### PR TITLE
Fpbg3 101 implement products collections & filters

### PR DESCRIPTION
**Pull request by rshgithub**
1. Get New Arrivals
Endpoint: `/products/new-arrivals`
Method: GET
Description: Retrieves products that have been created within the last 3 months.

2. Get Handpicked Collections
Endpoint: `/products/handpicked-collections`
Method: GET
Description: Retrieves handpicked products with a rating above 4.5 and a price less than $100.

3. Get Limited Edition Products
Endpoint:` /products/limited-edition`
Method: GET
Description: Retrieves products that have fewer than 20 units in stock.

4. Get Discounted Products
Endpoint:` /products/discounted-products`
Method: GET
Description: Retrieves products with a discount rate of 15% or more.

5. Get Popular Products
Endpoint: `/products/popular-products`
Method: GET
Description: Retrieves products with a rating of 4.5 or more.

6. Get Related Products
Endpoint: `/products/relatedProducts/{productName}`
Method: GET
Description: Retrieves up to 5 random products from the same category as the specified product.

7. Get Searched Products
Endpoint: `/products/search?term=example`
Method: GET
Description: Searches for products based on a search term in their name.

### CODE REVIEW

File: src/controllers/productsController.ts [hunk 1]
```diff
@@ -128,10 +129,162 @@ const updateProduct = errorHandler(
...
+      products: handpickedCollections.length > 0 ?
+        handpickedCollections : 'No handpicked collections found.',
+    });
+  },
+);
+
+// Fetch Limited Edition Products
+const getLimitedEditionProducts = errorHandler(
+  async(req: Request, res: Response, next: NextFunction) => {
+    const filterOptions = {
```
Comment (AhmadAlBarasy) [inline src/controllers/productsController.ts]: Why are you creating a filter for that while we already have a virtual field called `isLimitedEdition` that is a boolean value that you can check? please fix it, and also don't forget to fix the virtual field check evaluation in the `Products` model to only be true when the stock is less than 20 items, not 10.

File: src/controllers/productsController.ts [hunk 2]
```diff
@@ -128,10 +129,162 @@ const updateProduct = errorHandler(
...
+      discountRate: { [Op.gte]: 0.15 },
+    };
+
+    const discountedProducts = await productsService({}, undefined, filterOptions);
+
+    res.status(200).json({
+      status: 'success',
+      totalProducts: discountedProducts.length,
+      products: discountedProducts.length > 0 ?
+        discountedProducts : 'No products with 15% off or more found.',
```
Comment (AhmadAlBarasy) [inline src/controllers/productsController.ts]: Fix the message to be `'No products found with 15% discount rate or more'`

**Review by AhmadAlBarasy — CHANGES_REQUESTED** [review]
Please check the 2 small fixes I requested in the comments

File: src/controllers/productsController.ts (hunk 2 above)
Comment (rshgithub) [reply src/controllers/productsController.ts]: what the difference? both have the same meaning already .

File: src/controllers/productsController.ts (hunk 1 above)
Comment (rshgithub) [reply src/controllers/productsController.ts]: i missed that due to quickness .
done .

File: src/controllers/productsController.ts [hunk 3]
```diff
@@ -188,6 +189,148 @@ const deleteProductImage = errorHandler(
...
+      products: handpickedCollections.length > 0 ?
+        handpickedCollections : 'No handpicked collections found.',
+    });
+  },
+);
+
+const getLimitedEditionProducts = errorHandler(
+  async(req: Request, res: Response, next: NextFunction) => {
+    const allProducts = await productsService({}, undefined);
+    const limitedEditionProducts = allProducts.filter((product) => product.isLimitedEdition);
```
Comment (AhmadAlBarasy) [inline src/controllers/productsController.ts:235]: the filtering shouldn't be done at the code level, but on the SQL level, I'll approve for now.

**Review by AhmadAlBarasy — APPROVED** [review]

**Pull request merged**